secure/secure.sh

```diff
@@ -0,0 +1,403 @@
+    set_property_hive_site 'hive.metastore.kerberos.principal' "hive/_HOST@$REALM"
+    set_property_hive_site 'hive.metastore.kerberos.keytab.file' '/etc/hive/conf/hive.keytab'
+    # WebUI SSL
+    set_property_hive_site 'hive.server2.webui.use.ssl' 'true'
+    set_property_hive_site 'hive.server2.webui.keystore.path' "$HADOOP_CONF_DIR/ssl/keystore.jks"
+    set_property_hive_site 'hive.server2.webui.keystore.password' $keystore_password
+  fi
+}
+
+function secure_spark() {
+  if [[ $ROLE == 'Master' ]]; then
+    kadmin -p root -w $root_principal_password -q "addprinc -randkey spark/$FQDN"
+    kadmin -p root -w $root_principal_password -q "xst -k spark.keytab spark/$FQDN"
+    mv ./spark.keytab /etc/spark/conf/
+    chown spark:spark /etc/spark/conf/spark.keytab
+    chmod 400 /etc/spark/conf/spark.keytab
+    # Add the user 'spark' to group 'hadoop' so it can access the keystore files
+    usermod -a -G hadoop spark
+    cat << EOF >> /etc/spark/conf/spark-env.sh
+SPARK_HISTORY_OPTS="-Dspark.history.kerberos.enabled=true
+-Dspark.history.kerberos.principal=spark/$FQDN@$REALM
+-Dspark.history.kerberos.keytab=/etc/spark/conf/spark.keytab"
+EOF
+    cat << EOF >> /etc/spark/conf/spark-defaults.conf
+spark.yarn.historyServer.address https://$FQDN:18480
+spark.ssl.protocol tls
+spark.ssl.historyServer.enabled true
+spark.ssl.trustStore $HADOOP_CONF_DIR/ssl/truststore.jks
+spark.ssl.keyStore $HADOOP_CONF_DIR/ssl/keystore.jks
+spark.ssl.trustStorePassword $keystore_password
+spark.ssl.keyStorePassword $keystore_password
+EOF
+  fi
+}
+
+function mark_kerberized() {
+  touch /etc/google-dataproc/kerberized
+}
+
+function restart_services() {
+  if [[ $ROLE == 'Master' ]]; then
+    service hadoop-hdfs-namenode restart || err 'Cannot restart name node'
+    service hadoop-hdfs-secondarynamenode restart || err 'Cannot restart secondary name node'
+    service hadoop-mapreduce-historyserver restart || err 'Cannot restart mapreduce history server'
+    service hadoop-yarn-resourcemanager restart || err 'Cannot restart yarn resource manager'
+    service hive-server2 restart || err 'Cannot restart hive server'
+    service hive-metastore restart || err 'Cannot restart hive metastore'
+    service spark-history-server restart || err 'Cannot restart spark history server'
+  fi
+  # In single node mode, we have to run datanode and nodemanager on the master.
+  if [[ $ROLE == 'Worker' || $WORKER_COUNT == '0' ]]; then
+    service hadoop-hdfs-datanode restart || err 'Cannot restart data node'
+    service hadoop-yarn-nodemanager restart || err 'Cannot restart node manager'
+  fi
+}
+
+function main() {
+  update_apt_get
+  install_kerberos
+  install_jce
+  create_root_user_principal
+  create_service_principals
+  config_core_site
+  config_hdfs_site
+  config_yarn_site_and_permission
+  config_mapred_site
+  copy_keystore_files
+  config_ssl_xml 'server'
+  config_ssl_xml 'client'
+  secure_hive
+  secure_spark
+  restart_services
```

Comment:
In Dataproc 1.3 I think you'll need to kerberize the app timeline server as well :/

Reply:
Done.